iree/compiler/Codegen/Common/BufferizeCopyOnlyDispatchesPass.cpp

```diff
@@ -0,0 +1,247 @@
+  addToResult(lhs);
+  addToResult(rhs);
+  AffineMap map = AffineMap::get(0, operands.size(), resultExpr);
+  return builder.create<AffineApplyOp>(loc, map, operands).getResult();
+}
+
+/// Returns the offsets to use when combining two operations that implement the
+/// `OffsetSizeAndStrideOpInterface`. Also checks that the strides are 1.
+static FailureOr<SmallVector<OpFoldResult>> foldOffsetsSizesAndStrides(
+    OpBuilder &builder, Location loc, OffsetSizeAndStrideOpInterface producer,
+    OffsetSizeAndStrideOpInterface consumer) {
+  auto checkOne = [](OpFoldResult ofr) -> bool {
+    auto attr = ofr.dyn_cast<Attribute>();
+    return attr && attr.cast<IntegerAttr>().getInt() == 1;
+  };
+  auto producerStrides = producer.getMixedStrides();
+  auto consumerStrides = consumer.getMixedStrides();
+  if (producerStrides.size() != consumerStrides.size()) {
+    return static_cast<LogicalResult>(producer->emitOpError(
+        "expected same number of offsets/sizes/strides for producer and "
+        "consumer"));
+  }
+
+  if (!llvm::all_of(producer.getMixedStrides(), checkOne)) {
+    return static_cast<LogicalResult>(
+        producer->emitOpError("expected all strides to be 1"));
+  }
+  if (!llvm::all_of(consumer.getMixedStrides(), checkOne)) {
+    return static_cast<LogicalResult>(
+        consumer->emitOpError("expected all strides to be 1"));
+  }
+
+  // Combined offsets is the addition of the two offsets.
+  return llvm::to_vector(llvm::map_range(
+      llvm::zip(producer.getMixedOffsets(), consumer.getMixedOffsets()),
+      [&](std::tuple<OpFoldResult, OpFoldResult> t) {
+        return add(builder, loc, std::get<0>(t), std::get<1>(t));
+      }));
+}
+
+/// Returns true if the `v` is from a `flow.dispatch.tensor.load` operation.
+static bool isDirectlyFromDispatchTensorLoad(Value v) {
+  // Might eventually need to walk the use-def chain a bit, but for now,
+  // just check for the value defined by a flow.dispatch.tensor.load.
+  return v.getDefiningOp<IREE::Flow::DispatchTensorLoadOp>() != nullptr;
+}
+
+namespace {
+
+/// Pattern to fold `flow.dispatch.tensor.load` -> `tensor.extract_slice`.
+// TODO(ravishankarm): Eventually this should go in as a canonicalization at the
+// Flow level.
+struct FoldTensorLoadWithExtractSlice
+    : OpRewritePattern<tensor::ExtractSliceOp> {
+  using OpRewritePattern<tensor::ExtractSliceOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(tensor::ExtractSliceOp extractSliceOp,
+                                PatternRewriter &rewriter) const override {
+    auto dispatchTensorLoadOp =
+        extractSliceOp.source()
+            .getDefiningOp<IREE::Flow::DispatchTensorLoadOp>();
+    if (!dispatchTensorLoadOp) return failure();
+
+    FailureOr<SmallVector<OpFoldResult>> offsets =
+        foldOffsetsSizesAndStrides(rewriter, dispatchTensorLoadOp->getLoc(),
+                                   dispatchTensorLoadOp, extractSliceOp);
+    if (failed(offsets)) {
+      return failure();
```

Comment:
You probably want to emit the error here returned inside `offsets`?

Reply:
Hmm, I thought it will be dropped if not bubbled back to the driver. Maybe that was the old behavior..